### ISSUE TITLE
Bug - Adjusting Levels/XP

### ISSUE THREAD

**Issue by PassyFox** [P03W/RPGStats#25]
when i set the levels with the /rpgcheat command for defense it doesn't reset the amount of hearts people have to their name. Thus it keeps the hearts instead of recalculating how many they should have according to their set level. 

When trying to adjust the scoreboard value for "Hearts" it also says it's read only.

**Comment by P03W** [comment]
"Hearts" being readonly is a vanilla feature

Are you lowering their defense stat? Its possible the command doesnt take that into account

**Comment by PassyFox** [comment]
i "set" the "defense" "level" to a fixed value because of a trade the user made with me, i wanted them to pay in levels :P so i took 10 levels from him, he was 23 and set the level to 13, but he kept the hearts that various levels gave him.

**Comment by P03W** [comment]
Thanks, it should reset on death, but ill work on fixing that

**Comment by PassyFox** [comment]
i killed him a moment ago, but didn't reset. it's not too big of a problem and we'll gladly wait till it's done.

**Comment by P03W** [comment]
If it didnt reset on death thats a problem, because https://bugs.mojang.com/browse/MC-179940 is still open

Could you verify that his levels actually went down?

**Comment by P03W** [comment]
Cant verify health no resetting on death in dev, do you have any other mods on the server?

**Comment by PassyFox** [comment]
![Screenshot](https://user-images.githubusercontent.com/20614880/104817228-41bc7680-5820-11eb-8b39-bb535f52d92f.png)

it worked, i used the atribute command now and killed him and now it works, but the system didn't reset on it's own

**Comment by P03W** [comment]
Everything works for me in dev (other than the command not resetting automatically), could you send a modlist and maybe a log?

**Comment by PassyFox** [comment]
![modlist](https://user-images.githubusercontent.com/20614880/104817723-3e76ba00-5823-11eb-9727-ccaba2358fe2.PNG)

i used /kill to kill him now, but i meant the system doesn't reset when he dies normaly due to lava / player kill /  mob / fall/tnt damage

**Comment by P03W** [comment]
Yeah the issue is that there isn't any difference internally for those though

Thanks for the mod list ill check if one of them could be causing this

**Comment by PassyFox** [comment]
![2021-01-18_21 59 45](https://user-images.githubusercontent.com/20614880/104962702-44160080-59d9-11eb-9d14-a13e403be787.png)

it seems ur patch gives players negative XP in the defense stat, also using the /kill command now instantly gives so much xp they are lvl 50. somewhere it goes wrong.

might i suggest to keep the defense stat only based on the shield values (in statistics this is called "Damage Blocked by Shield") and not the rest of the armor or are those not splittable?

i understand the tought proces of having your armor being calculated as well, but seeying you need to difirentiate between what kind of damage people are getting i think having to use only the shield makes the effort of having more hearts also much more worth it seeyin this is an harder effort.

**Comment by P03W** [comment]
Only using the shield makes getting defense xp incredibly slow, as a full-face-on creeper explosion only gives maybe 50-80xp and can takeout a 3rd of shield

3.0.6 fixes infinite negative xp on /kill


Either I cant reproduce negative xp on falling or I fixed it with the kill fix as well

**Comment by PassyFox** [comment]
seems you fixed both :P

**Comment by PassyFox** [comment]
okay it seems peoaple are stil getting some negative XP, tested this myselfbut in the following cases people get negative xp: Walking on magma blocks, cactus, soul fire, campfire, wither rose or berry bushes
when people have armor but drop in lava they get positive xp added (with or without the fire protection), the same when they have featherfalling on boots (without the enchant nothing happens)
is their a way to seperate the damage source and perhaps have a setting which gives and which takes xp? or have a setting in the condig to only use shield (i know i have said this earlier but if you feal that goes to slow you can also adres the formula to give more xp per point damage taken to the shield?)

**Comment by P03W** [comment]
New version (3.1.0) should fix defence xp once and for all, as it caps it between 1 and 15

**Comment by P03W** [comment]
As for the original bug, after a bunch of thought (and staring at code) I don't see any reasonable way to fix this, especially with mod compatibility, so Ill be closing as wontfix, sorry

**Comment by PassyFox** [comment]
then a personal request from me and the fellow server members, is it possible you can show me the code so i can adjust it like it was in the beginning but it only uses the shield block value for calculations of the amount of xp gatherd?

**Comment by P03W** [comment]
Delete https://github.com/P03W/RPGStats/blob/master/src/main/java/mc/rpgstats/mixin/ApplyDamageMixin.java and remove `"ApplyDamageMixin",` from https://github.com/P03W/RPGStats/blob/master/src/main/resources/rpgstats.mixins.json
